data-products/src/main/scala/org/sunbird/analytics/audit/ScoreMetricMigrationJob.scala

```diff
@@ -0,0 +1,183 @@
+      val total_records = res._2.count()
+      JobLogger.log(s"Updating the $total_records records in the cassandra table", None, INFO)
+      updatedTable(res._2, userActivityAggDBSettings)
+      JobLogger.end(s"$jobName completed execution", "SUCCESS", Option(Map("timeTaken" -> res._1, "totalRecordsUpdated" -> total_records)))
+    } finally {
+      frameworkContext.closeContext()
+      spark.close()
+    }
+  }
+
+  // $COVERAGE-ON$
+
+  /**
+   * Get the migrateable dataframe by joining assessment_aggregator to user_activity_agg records
+   *
+   * Configs:
+   * batchId: List of batchids to migrate data for specific batch. if empty list, migrates data for all batches
+   * metricsType: type to store in agg_details as identifier of agg_detail
+   * forceMerge: option to forcefully replace in exisiting agg_details in user_activity_agg
+   *
+   * @param session SparkSession
+   * @param jobConfig JobConfig
+   *
+   * @return DataFrame
+   */
+  def migrateData(session: SparkSession, jobConfig: JobConfig): DataFrame = {
+    val batchIds: List[String] = jobConfig.modelParams.get.getOrElse("batchId", List()).asInstanceOf[List[String]]
+    val metricsType: String = jobConfig.modelParams.get.getOrElse("metricsType", "attempt_metrics").toString
+    val forceMerge: Boolean = jobConfig.modelParams.get.getOrElse("forceMerge", true).asInstanceOf[Boolean]
+
+    val updateAggColumn = udf(mergeAggMapCol())
+    val updatedAggLastUpdatedCol = udf(mergeAggLastUpdatedMapCol())
+    val flatAggList = udf(mergeAggListValues())
+    val flatAggLastUpdatedList = udf(mergeAggLastUpdatedListValues())
+    val flatAggDetailList = udf(mergeAggDetailListValues())
+    val updatedAggDetailsCol = udf(mergeAggDetailsCol())
+
+    val activityAggDF = fetchActivityData(session, batchIds)
+    val assessmentAggDF = getBestScoreAggDetailsDF(fetchAssessmentData(session, batchIds), metricsType)
+
+    val filterDF = activityAggDF.join(assessmentAggDF, activityAggDF.col("activity_id") === assessmentAggDF.col("course_id") &&
+      activityAggDF.col("userid") === assessmentAggDF.col("user_id") &&
+      activityAggDF.col("context_id") === assessmentAggDF.col("batchid"), joinType = "inner")
+      .select("agg", "aggregates", "agg_last_updated", "activity_type", "user_id", "context_id", "activity_id", "total_max_score", "best_score", "content_id", "agg_details", "migrating_agg_details", "attempts_count")
+
+    filterDF.withColumn("aggregates", updateAggColumn(when(lit(forceMerge), col("agg").cast("map<string, double>")).otherwise(col("aggregates").cast("map<string, double>")), col("total_max_score").cast(sql.types.DoubleType), col("best_score").cast(sql.types.DoubleType), col("attempts_count").cast(sql.types.DoubleType), col("content_id").cast(sql.types.StringType)))
+      .withColumn("agg_details", when(lit(forceMerge), col("migrating_agg_details")).otherwise(updatedAggDetailsCol(col("agg_details"), col("migrating_agg_details"))))
```

Comment:
The forceMerge logic can be moved to the UDF. This code would be much simpler to read

Reply:
Santhosh, We already have a udf function calling in this code. 
Creating nested udf is creating some issue. Could we please take up this review comment in next release ?